impl_core/src/main/java/io/opencensus/implcore/metrics/DerivedDoubleGaugeImpl.java

```
@@ -0,0 +1,137 @@
+      Maps.newHashMap();
+  private final int labelKeysSize;
+
+  DerivedDoubleGaugeImpl(String name, String description, String unit, List<LabelKey> labelKeys) {
+    labelKeysSize = labelKeys.size();
+    this.metricDescriptor =
+        MetricDescriptor.create(name, description, unit, Type.GAUGE_DOUBLE, labelKeys);
+  }
+
+  @Override
+  public <T> void createTimeSeries(
+      List<LabelValue> labelValues, @Nullable T obj, ToDoubleFunction<T> function) {
+    checkNotNull(labelValues, "labelValues should not be null.");
+    checkArgument(labelKeysSize == labelValues.size(), "Incorrect number of labels.");
+    Utils.checkListElementNotNull(labelValues, "labelValues element should not be null.");
+
+    registerTimeSeries(
+        new ArrayList<LabelValue>(labelValues), obj, checkNotNull(function, "function"));
+  }
+
+  @Override
+  public void removeTimeSeries(List<LabelValue> labelValues) {
+    checkNotNull(labelValues, "labelValues should not be null.");
+    Utils.checkListElementNotNull(labelValues, "labelValues element should not be null.");
+
+    remove(labelValues);
+  }
+
+  @Override
+  public synchronized void clear() {
+    registeredTimeSeries.clear();
+  }
+
+  private synchronized void remove(List<LabelValue> labelValues) {
+    Map<List<LabelValue>, TimeSeriesProducer> registeredTimeSeriesCopy =
+        new HashMap<List<LabelValue>, TimeSeriesProducer>(registeredTimeSeries);
+    if (registeredTimeSeriesCopy.remove(labelValues) == null) {
+      // The element not present, no need to update the current map of time series.
+      return;
+    }
+    registeredTimeSeries = Collections.unmodifiableMap(registeredTimeSeriesCopy);
+  }
+
+  private synchronized <T> void registerTimeSeries(
+      List<LabelValue> labelValues, @Nullable T obj, ToDoubleFunction<T> function) {
+    TimeSeriesProducer existingTimeSeries = registeredTimeSeries.get(labelValues);
+    if (existingTimeSeries != null) {
+      return;
```

Better to throw an exception or override the previous value. Just completely ignoring this will not signal the user.

---

Throwing an exception if already present.